### ISSUE TITLE
Feature: DHCP

### ISSUE THREAD

**Issue by gronke** [iocage/libiocage#15]
Jails should be allowed to set `ip4_addr=vnet0|dhcp` to obtain an IPv4 address via DHCP on the target interface `vnet0`.

DHCP requires access to /dev/bpf, which needs to be granted via udev rule on the host.

## Requirements
An acceptable solution:
- [ ] creates a set of udev rules per jail
- [ ] removes jail udev rules when stopping a jail
- when `dhcp` is set on any interface in ip4_addr
  - [ ] enables BPF in the jail's udev rule before starting the ip4_addr
- when `dhcp` is not set on any interface in ip4_addr
  - [ ] disables BPF in the jail's udev rule before starting the jail

**Comment by skarekrow** [comment]
This is overkill, and differs from the implementation in iocage currently. The current implementation uses `dhcp` and `bpf` properties as some users wanted to only enable bpf, but not use dhcp. So I think we should migrate any `dhcp` users to the supplied `ip4_addr` syntax here, and add a `bpf` property that is `on` when dhcp is present.

There is no reason to have a udev rule per jail for this however. All jails use the ruleset 4 currently, which is sufficient. We can create just one additional ruleset (exactly how iocage does it now) and have any DHCP jails use that. There isn't a reason to remove things from the devfs rules, that's unnecessary writing to the HDD that we can avoid.

**Comment by gronke** [comment]
> properties as some users wanted to only enable bpf, but not use dhcp

That would be another feature. Indeed it affects the requirements of this one, but it should be tackled in a separate ticket.

> There is no reason to have a udev rule per jail for this however

Maybe a jail want to combine udev rules. It's better to have tailored rules than opening up permissions across all jails.

> that's unnecessary writing to the HDD that we can avoid

Compared to the other operations involved during startup this seems marginal and there are multiple opportunities to optimize this in case it becomes a bottleneck. For example storing all possible combinations that can be activated and giving each a separate id.

**Comment by skarekrow** [comment]
> Maybe a jail want to combine udev rules. It's better to have tailored rules than opening up permissions across all jails.

That doesn't work how you expect. Stacking rules are ignored per the manpage:
```
include ruleset
[...]
Include commands in the referenced ruleset are not resolved.
```

@rexpo sent me a fantastic email detailing this, you can find a forum thread about it here: https://forums.freebsd.org/threads/46091/

That's why we need to stick to defined rulesets, and if a user wants his own he is responsible for making one and supplying that to iocage. We don't want to be responsible for a wide open `/dev` inside the jail.

**Comment by rexpo** [comment]
@gronke your `ip4_addr="vnet0|dhcp"` syntax is what I was also hoping for.

Especially in the case where I want a really cool jail set up like this:

```
ip4_addr="vnet0|dhcp,vnet1|192.168.1.12/29,vnet2|dhcp,vnet3|dhcp"
```

As for the `bpf` stuff:

> That's why we need to stick to defined rulesets, and if a user wants his own he is responsible for making one and supplying that to iocage.

Yeah - default to the Right Way to do it, but please don't take away all our opportunities to shoot ourselves in the foot if we want to, ok? :stuck_out_tongue_winking_eye:

Seriously though, user-supplied rulesets are needed for some situations, nice to have the option :)

**Comment by gronke** [comment]
> Seriously though, user-supplied rulesets are needed for some situations, nice to have the option :)

I agree. User defined overrides should always have preference. We should print a warning if the user selects a ruleset that does not have BPF enabled but used DHCP.

**Comment by rexpo** [comment]
> We should print a warning if the user selects a ruleset that does not have BPF enabled but used DHCP.

Yeah, that would be a nice touch. (It would require that you analyse the provided ruleset in `/etc/devfs.rules`, but that wouldn't be too difficult to implement)

**Comment by skarekrow** [comment]
I think both of you are misunderstanding me :)

I'm saying that combining udev rules as suggested by @gronke will lead us with a wide open `/dev`. I don't intend for any behavioral change from the current implementation.

That is `devfs_ruleset` can be any number, but besides the number `5`, which we use for DHCP enabled jails, the user is on their own. iocage will not construct a number `6` that is a blend of `1` and `3` for instance. That is best left to the user to create and supply `devfs_ruleset=6` to `iocage`

**Comment by rexpo** [comment]
Yeah we don't want a wide open `/dev` tree.

By foot-shooting, I was referring to a user-supplied ruleset 6 that doesn't unhide `bpf` on a DHCP-enabled jail. (getting a warning message printed is nice in this case)

**Comment by skarekrow** [comment]
Ah, in that case that already exists in the current implementation (https://github.com/iocage/iocage/blob/master/iocage/lib/ioc_start.py#L247-L254), there are a lot of things already in that code base ;)

**Comment by gronke** [comment]
> I'm saying that combining udev rules as suggested by @gronke will lead us with a wide open /dev

My suggestion was to use one rule per jail that defined the minimum amount of enabled actions. Of course the user may override that and directly reference a rule of choice.

**Comment by skarekrow** [comment]
> My suggestion was to use one rule per jail that defined the minimum amount of enabled actions. Of course the user may override that and directly reference a rule of choice.

We already do that with `devfs_ruleset=4`, or am I misunderstanding your suggestion?
